flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java

```diff
@@ -906,6 +919,256 @@ public void testChannelSelectors() {
+				new POJOwithHashCode(new int[] {1, 2}));
+
+		input.keyBy(new KeySelector<POJOwithHashCode, POJOwithHashCode>() {
+			@Override
+			public POJOwithHashCode getKey(POJOwithHashCode value) throws Exception {
+				return value;
+			}
+		}).addSink(new SinkFunction<POJOwithHashCode>() {
+			@Override
+			public void invoke(POJOwithHashCode value) throws Exception {
+				Assert.assertEquals(value.getId(), new int[]{1, 2});
+			}
+		});
+	}
+
+	@Test
+	public void testPOJOnoHashCodeKeyRejection() {
+
+		KeySelector<POJOnoHashCode, POJOnoHashCode> keySelector =
+				new KeySelector<POJOnoHashCode, POJOnoHashCode>() {
+					@Override
+					public POJOnoHashCode getKey(POJOnoHashCode value) throws Exception {
+						return value;
+					}
+				};
+
+		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+
+		DataStream<POJOnoHashCode> input = env.fromElements(
+				new POJOnoHashCode(new int[] {1, 2}));
+
+		// adjust the rule
+		expectedException.expect(InvalidProgramException.class);
+
+		input.keyBy(keySelector);
+	}
+
+	// composite key tests : Tuples
+
+	@Test
+	public void testTupleNestedArrayKeyRejection() {
+		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+
+		DataStream<Tuple2<Integer[], String>> input = env.fromElements(
+				new Tuple2<>(new Integer[] {1, 2}, "test-test"));
+
+		TypeInformation<?> expectedTypeInfo = new TupleTypeInfo<Tuple2<Integer[], String>>(
+				BasicArrayTypeInfo.INT_ARRAY_TYPE_INFO, BasicTypeInfo.STRING_TYPE_INFO);
+
+		// adjust the rule
+		expectedException.expect(InvalidProgramException.class);
+		expectedException.expectMessage("This type (" + expectedTypeInfo + ") cannot be used as key.");
+
+		input.keyBy(new KeySelector<Tuple2<Integer[],String>, Tuple2<Integer[],String>>() {
+			@Override
+			public Tuple2<Integer[], String> getKey(Tuple2<Integer[], String> value) throws Exception {
+				return value;
+			}
+		});
+	}
+
+	@Test
+	public void testPrimitiveKeyRejection() throws Exception {
+		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+		env.setParallelism(1);
+		env.setMaxParallelism(1);
+
+		DataStream<Integer> input = env.fromElements(new Integer(10000));
+
+		TypeInformation<?> expectedTypeInfo = IntegerTypeInfo.INT_TYPE_INFO;
+
+		KeyedStream<Integer, Integer> keyedStream = input.keyBy(new KeySelector<Integer, Integer>() {
+			@Override
+			public Integer getKey(Integer value) throws Exception {
+				return value;
+			}
+		});
+
+		Assert.assertEquals(expectedTypeInfo, keyedStream.getKeyType());
+
+		keyedStream.addSink(new SinkFunction<Integer>() {
+			@Override
+			public void invoke(Integer value) throws Exception {
+				Assert.assertEquals(10000L, (long) value);
+			}
+		});
+
+		env.execute();
+	}
+
+	private static class TestClass {
+
+		private final int id;
+
+		TestClass(int id) {
+			this.id = id;
+		}
+	}
+
+	public static class POJOnoHashCode {
```

Comment:
the naming is also inconsistent with the other class; this class name doesn't contain "with".